pkg/placement/ha_test.go

```diff
@@ -210,81 +240,111 @@ func TestPlacementHA(t *testing.T) {
 }
 
 func createRaftServer(t *testing.T, nodeID int, peers []raft.PeerInfo) (*raft.Server, <-chan struct{}, context.CancelFunc) {
-	srv := raft.New(fmt.Sprintf("mynode-%d", nodeID), true, peers, "")
+	t.Helper()
+	clock := clocktesting.NewFakeClock(time.Now())
+
+	srv := raft.New(raft.Options{
+		ID:           fmt.Sprintf("mynode-%d", nodeID),
+		InMem:        true,
+		Peers:        peers,
+		LogStorePath: "",
+		Clock:        clock,
+	})
 
 	ctx, cancel := context.WithCancel(context.Background())
-	serverStopped := make(chan struct{})
+
+	stopped := make(chan struct{})
 	go func() {
-		defer close(serverStopped)
+		defer close(stopped)
 		require.NoError(t, srv.StartRaft(ctx, &hcraft.Config{
 			ProtocolVersion:    hcraft.ProtocolVersionMax,
-			HeartbeatTimeout:   250 * time.Millisecond,
-			ElectionTimeout:    250 * time.Millisecond,
-			CommitTimeout:      50 * time.Millisecond,
+			HeartbeatTimeout:   2 * time.Second,
+			ElectionTimeout:    2 * time.Second,
+			CommitTimeout:      2 * time.Second,
 			MaxAppendEntries:   64,
 			ShutdownOnRemove:   true,
 			TrailingLogs:       10240,
 			SnapshotInterval:   120 * time.Second,
 			SnapshotThreshold:  8192,
-			LeaderLeaseTimeout: 250 * time.Millisecond,
+			LeaderLeaseTimeout: time.Second,
+			BatchApplyCh:       true,
 		}))
 	}()
 
 	ready := make(chan struct{})
 	go func() {
 		defer close(ready)
-		timeoutCtx, timeoutCancel := context.WithTimeout(ctx, 20*time.Second)
-		defer timeoutCancel()
-		_, err := srv.Raft(timeoutCtx)
-		require.NoError(t, err)
+		for {
+			timeoutCtx, timeoutCancel := context.WithTimeout(ctx, time.Second*5)
+			defer timeoutCancel()
+			r, err := srv.Raft(timeoutCtx)
+			require.NoError(t, err)
+			if r.State() == hcraft.Follower || r.State() == hcraft.Leader {
+				return
+			}
+		}
+	}()
+
+	go func() {
+		for {
+			select {
+			case <-ready:
+			case <-time.After(time.Millisecond):
+				clock.Step(time.Second * 2)
+			}
+		}
 	}()
 
-	stopFn := func() {
+	return srv, ready, func() {
 		cancel()
 		select {
-		case <-serverStopped:
-			// nop
-		case <-time.After(5 * time.Second):
-			t.Fatal("raft server did not stop in time")
+		case <-stopped:
+		case <-time.After(time.Second * 5):
+			require.Fail(t, "server didn't stop in time")
 		}
 	}
-
-	return srv, ready, stopFn
 }
 
 func findLeader(t *testing.T, raftServers []*raft.Server) int {
-	n := atomic.Int32{}
-	n.Store(-1)
+	t.Helper()
```

Comment:
I recommend not marking functions as test helpers if they contain assertions inside.

When a function is a helper, a failed test inside will be reported in the line of the caller. See docs for Helper(): https://pkg.go.dev/testing#T.Helper


```suggestion
```

Reply:
:+1: